rust_dev_preview/examples/rds/src/aurora.rs

```diff
@@ -0,0 +1,664 @@
+    types::{DbCluster, DbClusterParameterGroup, DbClusterSnapshot, DbInstance, Parameter},
+    Client,
+};
+use tracing::{info, trace, warn};
+
+const DB_ENGINE: &str = "aurora-mysql";
+const DB_CLUSTER_PARAMETER_GROUP_NAME: &str = "RustSDKCodeExamplesDBParameterGroup";
+const DB_CLUSTER_PARAMETER_GROUP_DESCRIPTION: &str =
+    "Parameter Group created by Rust SDK Code Example";
+const DB_CLUSTER_IDENTIFIER: &str = "RustSDKCodeExamplesDBCluster";
+const DB_INSTANCE_IDENTIFIER: &str = "RustSDKCodeExamplesDBInstance";
+
+const MAX_WAIT: Duration = Duration::from_secs(5 * 60); // Wait at most 25 seconds.
+const WAIT_TIME: Duration = Duration::from_millis(500); // Wait half a second at a time.
+static FILTER_PARAMETER_NAMES: Set<&'static str> = phf_set! {
+    "auto_increment_offset",
+    "auto_increment_increment",
+};
+
+struct Waiter {
+    start: SystemTime,
+    max: Duration,
+}
+
+impl Waiter {
+    fn new(max: Duration) -> Self {
+        Waiter {
+            start: SystemTime::now(),
+            max,
+        }
+    }
+
+    async fn sleep(&self) -> Result<(), ScenarioError> {
+        if SystemTime::now()
+            .duration_since(self.start)
+            .unwrap_or(Duration::MAX)
+            > self.max
+        {
+            Err(ScenarioError::with(
+                "Exceeded maximum wait duration for stable group",
+            ))
+        } else {
+            tokio::time::sleep(WAIT_TIME).await;
+            Ok(())
+        }
+    }
+}
+
+impl Default for Waiter {
+    fn default() -> Self {
+        Self::new(MAX_WAIT)
+    }
+}
+
+#[derive(Debug)]
+struct MetadataError {
+    message: Option<String>,
+    code: Option<String>,
+}
+
+impl MetadataError {
+    fn from(err: &dyn ProvideErrorMetadata) -> Self {
+        MetadataError {
+            message: err.message().map(|s| s.to_string()),
+            code: err.code().map(|s| s.to_string()),
+        }
+    }
+}
+
+impl Display for MetadataError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let display = match (&self.message, &self.code) {
+            (None, None) => "Unknown".to_string(),
+            (None, Some(code)) => format!("({code})"),
+            (Some(message), None) => message.to_string(),
+            (Some(message), Some(code)) => format!("{message} ({code})"),
+        };
+        write!(f, "{display}")
+    }
+}
+
+#[derive(Debug)]
+pub struct ScenarioError {
+    message: String,
+    context: Option<MetadataError>,
+}
+
+impl ScenarioError {
+    pub fn with(message: impl Into<String>) -> Self {
+        ScenarioError {
+            message: message.into(),
+            context: None,
+        }
+    }
+
+    pub fn new(message: impl Into<String>, err: &dyn ProvideErrorMetadata) -> Self {
+        ScenarioError {
+            message: message.into(),
+            context: Some(MetadataError::from(err)),
+        }
+    }
+}
+
+impl std::error::Error for ScenarioError {}
```

Comment:
What's this for?

Reply:
This tells the type system that ScenarioError is an Error. In Java, this would be `ScenarioError implements Error`